### PR TITLE
Fix UVM compatibility with CUDA 13.

### PR DESCRIPTION
**Pull request by lmcafee-nvidia**
# What does this PR do ?
Fix UVM (unified virtual memory) compatibility with CUDA 13. The function signature for `cudaMemAdvise()` changed from CUDA 12 to 13, and this PR handles this change.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share discuss a design-doc with the team.  

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/inference/unified_memory.py [hunk 1]
```diff
@@ -48,6 +49,33 @@ def compile_allocator():
     if _compilation_state != CompilationState.UNATTEMPTED:
         return
 
+    # Specialize `cudaMemAdvise` args for different CUDA versions.
+    cuda_version = float(torch.version.cuda)
```
Comment (tdene) [inline megatron/core/inference/unified_memory.py]: After thinking about this more carefully, I do not think we should check the version in Python. I think we should check it in C instead.

```
      if (device >= 0) {
        // cudaMemAdviseSetPreferredLocation sets the preferred location for the memory.
        // This is a hint that tries to prevent data from being migrated away from the device.

        #if CUDART_VERSION >= 13000
          // For CUDA >= 13, the cudaMemAdvise device arg is type cudaMemLocation
          // instead of an int, so we setup the location and conditionally use it
          // in calls to cudaMemAdvise.
          cudaMemLocation location = {0};
          location.type = cudaMemLocationTypeDevice;
          location.id = device;

          cudaMemAdvise(ptr, (size_t)size, cudaMemAdviseSetPreferredLocation, location);

          // cudaMemAdviseSetAccessedBy ensures the memory always lives in the device's page table.
          // Even if the memory has to be migrated away from the device, it still does not page fault.
          // The CUDA docs claim that cudaMemAdviseSetPreferredLocation completely overrides this flag,
          // but there is no harm in adding this flag as well for future-proofing.
          cudaMemAdvise(ptr, (size_t)size, cudaMemAdviseSetAccessedBy, location);
        #else
          // For CUDA < 13, cudaMemAdvise takes an int device argument.
          cudaMemAdvise(ptr, (size_t)size, cudaMemAdviseSetPreferredLocation, device);

          // cudaMemAdviseSetAccessedBy ensures the memory always lives in the device's page table.
          cudaMemAdvise(ptr, (size_t)size, cudaMemAdviseSetAccessedBy, device);
        #endif
      }
```

File: megatron/core/inference/unified_memory.py [hunk 2]
```diff
@@ -1,6 +1,7 @@
 # Copyright (c) 2025, NVIDIA CORPORATION. All rights reserved.
 
 import os
+import torch
```
Comment (tdene) [inline megatron/core/inference/unified_memory.py]: I'm putting this comment on an unrelated diff, because this is about the pre-existing code.

We do `cudaSetDevice` but we never change it back. This sounds bad. I think the following should be added before `return ptr;` inside `managed_malloc`:
```
if (device != cur && device >= 0) cudaSetDevice(cur);
```

File: megatron/core/inference/unified_memory.py (hunk 1 above)
Comment (lmcafee-nvidia) [reply megatron/core/inference/unified_memory.py]: I initially tried doing this in C but ran into type errors, but I wasn't using `#` directives like you have. Let me try this way and see if it works.

Comment (tdene) [reply megatron/core/inference/unified_memory.py]: The reason why I bring this up is because I started looking through the pytorch forums about whether it's guaranteed that `torch.version.cuda` will actually match `cuda_runtime_api.h`, and corner-cases, and it just seems a lot safer to keep the type-handling as close to the source-of-truth as possible.

Comment (lmcafee-nvidia) [reply megatron/core/inference/unified_memory.py]: just fyi, it didn't like this line:
```
/root/.cache/torch_extensions/py312_cu130/managed_alloc_runtime/main.cpp: In function ‘void* managed_malloc(size_t, int, void*)’:
/root/.cache/torch_extensions/py312_cu130/managed_alloc_runtime/main.cpp:28:39: error: invalid conversion from ‘int’ to ‘cudaMemLocationType’ [-fpermissive]
   28 |           cudaMemLocation location = {0};
      |                                       ^
      |                                       |
      |                                       int
ninja: build stopped: subcommand failed.
```

so I'm leaving it as `cudaMemLocation location;`. Let me know if want to change it to something else.

**Review by kvareddy — APPROVED** [review]

**Comment by kvareddy** [comment]
@shanmugamr1992  can you please take a loot at this MR?
